Add use of `|` and `|=` operators to `Table`

AlinderS opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

Since Python 3.9, dictionaries support using the `|` and `|=` operators to merge and update dictionaries in a similar way to how `Table.update()` works. The goal of this pull request is to enable the use of `|` and `|=` operators for merge and update in place for Tables. No real new functionality is added, everything builds on the existing `.update` method. This PR only enables the operators. 

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

taldcroft commented:
Nice, thanks @AlinderS ! 

One more thing would be to add some mention in the sphinx narrative docs and also in the `update()` method docstring, noting this alternative syntax that is available. This would be in the "Perform a dictionary-style update" section of https://docs.astropy.org/en/stable/table/modify_table.html.

The sphinx docs seem to be failing for some reason, probably unrelated. If you didn't start on latest main then you should rebase on main, otherwise sometimes these things just clear up.

taldcroft commented on docs/changes/table/14187.feature.rst:
```diff
@@ -0,0 +1,2 @@
+Astropy table can now use the | and |= operators to perform dictionary-style
```
```suggestion
Astropy table can now use the ``|`` and ``|=`` operators to perform dictionary-style
```

taldcroft commented on astropy/table/table.py:
```diff
@@ -2262,6 +2262,21 @@ def __len__(self):
+            updated_table = self.copy()
+            updated_table.update(other, copy=True)
+            return updated_table
+        except TypeError:
```
You should test these exceptions.

AlinderS replied:
Sorry (first time doing this so I want to be super clear), do you mean I should update the tests to run code we expect to cause `TypeError`s and check that it returns `NotImplemented`?

taldcroft replied:
Yup!

taldcroft replied:
Just playing with this gives what might be a surprising result to some users. The result for `1 | t` makes sense, but then `t | 1` gives a `ValueError` related to not being able to create a Table with `1`.

In retrospect having the `update()` method auto-create a Table from the input seems like not the best idea, but we're stuck with the existing API for that method. 

BUT, for the new `|` and `|=` operators, maybe we should insist on `other` being a `Table` subclass and raise `TypeError` if not.

```
In [14]: 1 | t
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [14], in <module>
----> 1 1 | t

TypeError: unsupported operand type(s) for |: 'int' and 'Table'

In [15]: t | 1
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [15], in <module>
----> 1 t | 1

File ~/git/astropy/astropy/table/table.py:2268, in Table.__or__(self, other)
   2266 try:
   2267     updated_table = self.copy()
-> 2268     updated_table.update(other, copy=True)
   2269     return updated_table
   2270 except TypeError:

File ~/git/astropy/astropy/table/table.py:3413, in Table.update(self, other, copy)
   3410 from .operations import _merge_table_meta
   3412 if not isinstance(other, Table):
-> 3413     other = self.__class__(other, copy=copy)
   3414 common_cols = set(self.colnames).intersection(other.colnames)
   3415 for name, col in other.items():

File ~/git/astropy/astropy/table/table.py:852, in Table.__init__(self, data, masked, names, dtype, meta, copy, rows, copy_indices, units, descriptions, **kwargs)
    849     data = [[]] * n_cols
    851 else:
--> 852     raise ValueError(f"Data type {type(data)} not allowed to init Table")
    854 # Set up defaults if names and/or dtype are not specified.
    855 # A value of None means the actual value will be inferred
    856 # within the appropriate initialization routine, either from
    857 # existing specification or auto-generated.
    859 if dtype is None:

ValueError: Data type <class 'int'> not allowed to init Table
```

mhvk replied:
I agree with testing for `Table` here, i.e., something like
```
if not isinstance(other, Table):
    return NotImplemented
...
```
With that, any errors should just be raised, so no need for `try/except`.

AlinderS commented:
Do you think we should expand any of the examples to include demonstrations of the new notation?

nstarman commented:
> Do you think we should expand any of the examples to include demonstrations of the new notation?

Yes. Personally, I also think many of the examples should be changed to ``|=``. The new syntax is cleaner and I think built-in operators should be emphasized over methods, esp in the docs because methods are more easily discoverable than operators when working in an IDE.

taldcroft commented:
> Yes. Personally, I also think many of the examples should be changed to |=. 

Agreed.

pllim commented:
Probably needs a what's new when this is ready to be wrapped up? Thanks!

mhvk commented on astropy/table/table.py:
```diff
@@ -2262,6 +2262,21 @@ def __len__(self):
+    def __or__(self, other):
+        try:
+            updated_table = self.copy()
+            updated_table.update(other, copy=True)
```
For `dict` updates, no copies are made. Am happy to go with the default for `.update()` here (but would then remove `copy=True` here), since perhaps it is too risky otherwise. But thought I would mention it, since it is also possible the main reason for `copy=True` is that when tables were based on structured arrays, there really wasn't an option, and that we wanted to be backward-compatible. But this is a new operator, so we can decide differently.

mhvk commented:
p.s. Very nice addition!

AlinderS commented:
I think the last commit addresses the points above, please let me know if there is anything else that needs to be done. 
If not, that only leaves the "what's new" entry. Do I need to make it? I could not find how in the docs.

mhvk commented:
Looks good! You should add a changelog fragment in `docs/table/14187.feature.rst` (see `docs/changes/README.rst`). 

An additional what's-new indeed seems warranted - add something to `docs/whatsnew/5.3.rst` (being the first makes this perhaps a little trickier; see the current what's new web page for 5.2 to get a sense).

eerovaher commented on astropy/table/table.py:
```diff
@@ -3349,7 +3364,10 @@ def update(self, other, copy=True):
         to initialize a table. Columns from (possibly converted) ``other`` are
         added to this table. In case of matching column names the column from
-        this table is replaced with the one from ``other``.
+        this table is replaced with the one from ``other``. The operator
+        ``|=`` is available as alternate syntax for in-place update and ``|``
+        is available for assigning the updated table to a new table. These operators
+        require ``other`` to be a |Table| subclass.
```
```suggestion
        this table is replaced with the one from ``other``. If ``other`` is a
        |Table| instance then ``|=`` is available as alternate syntax for in-place
        update and ``|`` can be used merge data to a new table.
```

eerovaher commented on docs/table/modify_table.rst:
```diff
@@ -164,14 +173,18 @@ the table and replaces existing ones::
-The input of :meth:`~astropy.table.Table.update` does not have to be a |Table|,
-it can be anything that can be used for :ref:`construct_table` with a
-compatible number of rows.
+The ``|`` and  ``|=`` operators are built on the
+:meth:`~astropy.table.Table.update` method, but the
```
This is an implementation detail that should not be mentioned in the user documentation.

eerovaher reviewed (commented):
In case of a Python `dict` `|=` is less strict about the type of `other` than `|`:
```python
>>> a = {"a": 1}
>>> a | [["b", 2]]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for |: 'dict' and 'list'
>>> a |= [["b", 2]]
>>> a
{'a': 1, 'b': 2}
```
Restricting `|` to pairs of `Table` instances is consistent with that, but consistency with `dict` would suggest being less restrictive with `|=`.

nstarman commented on docs/whatsnew/5.3.rst:
```diff
@@ -22,6 +22,40 @@ By the numbers:
+	>>> from astropy.table import Table
+	>>> t1 = Table({'a': ['foo', 'bar'], 'b': [0., 0.]}, meta={'i': 0})
+	>>> t2 = Table({'b': [1., 2.], 'c': [7., 11.]}, meta={'n': 2})
+	t3 = t1 | t2  # Create new table as result of update
```
```suggestion
	>>> t3 = t1 | t2  # Create new table as result of update
```

AlinderS commented:
I think I have resolved the latest round of feedback. The `|=` now just tries to do `update()` instead of checking the type of `other`. I have edited the whats-new file, do you think it looks ok?
The read the Docs build still fails and I don't understand why. Do I need to do something to address this? Do you have any other comments?

nstarman commented:
Try rebasing on main, I think that should resolve the issue.

nstarman commented on astropy/table/table.py:
```diff
@@ -2262,6 +2262,21 @@ def __len__(self):
+            updated_table.update(other)
+            return updated_table
+        else:
+            return NotImplemented
```
Adding a test for this line and the one in `ior`` should fix the coverage failures! Then this PR is probably ready to be merged. Thanks @AlinderS

mhvk replied:
Note to the comment: How I would test this is by its effect, i.e., that `some_table | some_dict` gives a `TypeError`

taldcroft replied:
I don't understand this coverage report since the tests actually do run `t1 | 1` which I explicitly confirmed does run the `return NotImplemented` line. There is also test coverage for `t1 |= 1`. Any ideas? We can just ignore these coverage warnings but it is annoying to not understand.

mhvk replied:
Very weird.  And serves me right for not looking properly -- continuing in that vein nevertheless (no time to test myself...), if one runs `tox -e test-coverage` on one's own computer, does it also mark it as not covered?

Anyway, I would suggest to ignore this...

taldcroft commented on docs/whatsnew/5.3.rst:
```diff
@@ -22,6 +22,40 @@ By the numbers:
+New operators for Table updates
+===============================
+
+New operators for dictionary-style updates of |Table| are available. ``|=``
+works the same way as ``update()`` and updates the table in place. ``|`` return
+the updated table.
```
```suggestion
We have added support for dictionary-style merge (``|``) and update (``|=``) 
of columns in the |Table| class. This follows the  behavior for ``dict`` defined
in `https://peps.python.org/pep-0584/`_.
```

taldcroft commented on docs/whatsnew/5.3.rst:
```diff
@@ -22,6 +22,40 @@ By the numbers:
+the updated table.
+
+	>>> from astropy.table import Table
+	>>> t1 = Table({'a': ['foo', 'bar'], 'b': [0., 0.]}, meta={'i': 0})
+	>>> t2 = Table({'b': [1., 2.], 'c': [7., 11.]}, meta={'n': 2})
+	>>> t3 = t1 | t2  # Create new table as result of update
```
```suggestion
	>>> t1 = Table({'a': ['foo', 'bar'], 'b': [0., 0.]})
	>>> t2 = Table({'b': [1., 2.], 'c': [7., 11.]})
	>>> t3 = t1 | t2  # Create new table which merges columns from t1 and t2
```

taldcroft commented on docs/whatsnew/5.3.rst:
```diff
@@ -22,6 +22,40 @@ By the numbers:
+	foo     1.0     7.0
+	bar     2.0    11.0
+
+	>>> t1 |= (t2)
```
```suggestion
	>>> t1 |= (t2)  # Update t1 columns in-place with t2 columns
```

taldcroft commented on docs/whatsnew/5.3.rst:
```diff
@@ -22,6 +22,40 @@ By the numbers:
+	foo     1.0     7.0
+	bar     2.0    11.0
+
+``|`` only supports subclasses of |Table| whereas ``|=`` accepts any object that
+can be used to construct a table.
+
```
This is a detail more for the documentation than What's New.
```suggestion
```

taldcroft reviewed (commented):
I made a few suggestions for the What's New entry, otherwise this is looking great!

eerovaher commented on docs/table/modify_table.rst:
```diff
@@ -164,14 +164,33 @@ the table and replaces existing ones::
    bar     2.0    10.0
 
-:meth:`~astropy.table.Table.update` also takes care of silently :ref:`merging_metadata`::
+When using ``|=``, the other object does not need to be a |Table|, it can be
+anything that can be used for :ref:`construct_table` with a compatible number
+of rows.
```
It would be good to have a code example for updating with a `dict`.

eerovaher commented on docs/table/modify_table.rst:
```diff
@@ -164,14 +164,33 @@ the table and replaces existing ones::
+of rows.
 
-  >>> t1.meta
+It is also possible to use ``|`` to return the new |Table| rather than update
+in-place. ``|`` requires the other object to be a |Table|::
```
```suggestion
It is also possible to use the ``|`` operator to merge multiple |Table| instances
into a new table::
```
I would also suggest writing an example of attempting to merge a `Table` with something that's not a `Table`. If a user triggers this error and tries to find documentation about it then an explicit example with the correct error message might make it easier to find with a search engine.

eerovaher commented on docs/table/modify_table.rst:
```diff
@@ -164,14 +164,33 @@ the table and replaces existing ones::
+in-place. ``|`` requires the other object to be a |Table|::
+
+  >>> t1 = Table({'name': ['foo', 'bar'], 'val': [0., 0.]}, meta={'n': 2})
+  >>> t2 = Table({'val': [1., 2.], 'val2': [10., 10.]}, meta={'id': 0})
```
I'd recommend making `t2` a `QTable` and demonstrating merging both `t1 | t2` and `t2 | t1`. Such examples would document that the output has the same type as the leftmost table to be merged.
```suggestion
  >>> t2 = QTable({'val': [1., 2.], 'val2': [10., 10.]}, meta={'id': 0})
```

eerovaher commented on docs/table/modify_table.rst:
```diff
@@ -164,14 +164,33 @@ the table and replaces existing ones::
-compatible number of rows.
+The columns in the updated |Table| are going to be copies of the originals. If
+you need them to be references you can use the
+:meth:`~astropy.table.Table.update` method, see :ref:`copy_versus_reference`
```
```suggestion
:meth:`~astropy.table.Table.update` method with ``copy=False``, see :ref:`copy_versus_reference`
```

eerovaher commented on docs/whatsnew/5.3.rst:
```diff
@@ -22,6 +22,37 @@ By the numbers:
+
+.. _whatsnew-5.3-table-update-operators:
+
+Add union operators to Table
+===============================
```
```suggestion
Add union operators to |Table|
==============================
```

eerovaher commented on docs/whatsnew/5.3.rst:
```diff
@@ -22,6 +22,37 @@ By the numbers:
+	foo     1.0     7.0
+	bar     2.0    11.0
+
+	>>> t1 |= (t2)  # Update t1 columns in-place with t2 columns
```
```suggestion
	>>> t1 |= (t2)  # Update t1 in-place with t2
```

eerovaher commented on docs/changes/table/14187.feature.rst:
```diff
@@ -0,0 +1,2 @@
+Table and QTable can now use the ``|`` and ``|=`` operators to perform
+dictionary-style update and update in place.
```
PEP 584 calls `|` the merge operator and `|=` the update operator. I'd prefer to be consistent with those names.
```suggestion
``Table`` and ``QTable`` can now use the ``|`` and ``|=`` operators for
dictionary-style merge and update.
```

eerovaher commented on astropy/table/tests/test_table.py:
```diff
@@ -2455,6 +2455,42 @@ def test_update_without_copy(self):
         assert np.all(t2["c"] == self.c)
         assert np.all(t2["d"] == self.d)
 
+    def test_pipe_operator(self):
```
```suggestion
    def test_merge_operator(self):
```

eerovaher commented on astropy/table/tests/test_table.py:
```diff
@@ -2455,6 +2455,42 @@ def test_update_without_copy(self):
+        assert np.all(t3["b"] == t1_copy["b"])
+        assert np.all(t3["c"] == t1_copy["c"])
+
+    def test_pipe_inplace_operator(self):
```
```suggestion
    def test_update_operator(self):
```

eerovaher commented on docs/whatsnew/5.3.rst:
```diff
@@ -12,7 +12,7 @@ the 5.2 release.
 In particular, this release includes:
 
-.. * :ref:`whatsnew-5.3-example`
+.. * :ref:`whatsnew-5.3-table-update-operators`
```
```suggestion
.. * :ref:`whatsnew-5.3-table-union-operators`
```

eerovaher commented on docs/whatsnew/5.3.rst:
```diff
@@ -22,6 +22,37 @@ By the numbers:
 * X distinct people have contributed code
 
+
+.. _whatsnew-5.3-table-update-operators:
```
```suggestion
.. _whatsnew-5.3-table-union-operators:
```

eerovaher reviewed (commented):
I'd recommend expanding the documentation with a few more examples, and to be more consistent with how PEP 584 names the operators.

AlinderS commented:
The documentation is a bit nicer now. Assuming that we can ignore the codecov failures, and that the readthedocs test can be fixed with another rebase, is there anything more you think should be done before merging?

taldcroft commented:
@AlinderS - sorry for the delay. This is looking great now from the code / docs perspective, thanks for this nice feature!

So to get this across the finish line, please go ahead and rebase on main and fix the conflict (and hopefully the RTD error). It would also be good to squash down to one commit at this point.

taldcroft reviewed (approved):
Looks great, thanks @AlinderS !
